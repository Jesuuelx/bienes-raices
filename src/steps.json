[
  {
    "nameComponent": "NameStep",
    "name": "Nombre y Apellido",
    "pathTo": "/correo-usuario",
    "id:": 0,
    "description": "¿Quién recibirá la oferta?"
  },
  {
    "nameComponent": "EmailStep",
    "name": "Correo",
    "pathTo": "/direccion",
    "id:": 1,
    "description": "¿A que correo quieres que nos comuniquemos contigo?"
  },
  {
    "nameComponent": "AddressStep",
    "name": "Direccion",
    "pathTo": "/datos-piso",
    "id:": 2,
    "description": "Danos tu direccion"
  },
  {
    "nameComponent": "FloorStep",
    "name": "Piso",
    "pathTo": "/caracteristicas-adicional",
    "id:": 3,
    "description": "Piso de tu apartamento*"
  },
  {
    "nameComponent": "AdditionalStep",
    "name": "Informacion Adicional",
    "pathTo": "/resumen",
    "id:": 4,
    "description": "Elige si es tu caso"
  },
  {
    "nameComponent": "Resumen",
    "name": "Resumen",
    "id:": 5,
    "description": "¡Has completado tu registro con exito!"
  }
]
