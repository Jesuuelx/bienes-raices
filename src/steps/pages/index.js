export * from './AdditionalStep';
export * from './AddressStep';
export * from './EmailStep';
export * from './FloorStep';
export * from './NameStep';
