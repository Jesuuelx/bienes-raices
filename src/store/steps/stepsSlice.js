import { createSlice } from "@reduxjs/toolkit";

export const stepsSlice = createSlice({
  name: "steps",
  initialState: {
    status: "checking",
    displayName: null,
    email: null,
    address: null,
    floor: null,
    aditionalInformation: [
      {
        name: "zonabbq",
        done: false,
        id: 1,
      },
      {
        name: "parque",
        done: false,
        id: 2,
      },
      {
        name: "salondefiesta",
        done: false,
        id: 3,
      },
    ],
    isLoading: false,
    errorMessage: false,
  },
  reducers: {
    stepDisplayName: (state, { payload }) => {
      state.displayName = payload.name;
    },
    stepEmail: (state, { payload }) => {
      state.email = payload.email;
    },
    stepAddress: (state, { payload }) => {
      state.address = payload.address;
    },
    stepFloor: (state, { payload }) => {
      state.floor = payload.floor;
    },
    stepAditionalInformation: (state, { payload }) => {
      const newState = state.aditionalInformation.map((info) =>
        info.id === payload ? { ...info, done: !info.done } : info
      );
      state.aditionalInformation = newState;
    },
    startLoading: (state) => {
      state.isLoading = true;
    },
  },
});
export const {
  stepDisplayName,
  stepEmail,
  stepAddress,
  stepFloor,
  stepAditionalInformation,
  startLoading,
} = stepsSlice.actions;
