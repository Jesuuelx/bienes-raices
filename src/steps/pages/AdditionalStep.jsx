import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import { stepAditionalInformation } from "../../store/steps/stepsSlice";

export const AdditionalStep = () => {
  const dispatch = useDispatch();

  const navigate = useNavigate();
  
  const onSubmit = (e) => {
    e.preventDefault();
    
    navigate('/resumen', {replace:true});
  };

  return (
    <>
      <h1 className="step__title">
        Paso 5 de 5: <br />{" "}
        <span className="step__intro-title"> Opcional *</span>
      </h1>
      <h2 className="mb-10 step__solicita">Elige si es tu caso</h2>
      <hr />

      <form onSubmit={onSubmit} className="step__form-direccion">
        <div>
          <label className="mb-5"> Zona BBQ </label>
          <input
            type="checkbox"
            value="Zona BBQ"
            name="input"
            className="step__input-adicional"
            autoComplete="off"
            onChange={() => dispatch(stepAditionalInformation(1))}
          />
        </div>
        <br />
        <div>
          <label className="mb-5"> Salon Comunal </label>
          <input
            type="checkbox"
            value="Salon Comunal"
            name="input"
            className="step__input-adicional"
            autoComplete="off"
            onChange={() => dispatch(stepAditionalInformation(2))}
          />
        </div>
        <div>
          <label className="mb-5"> Parque de juegos </label>
          <input
            type="checkbox"
            value="Parque de juegos"
            name="input"
            className="step__input-adicional"
            autoComplete="off"
            onChange={() => dispatch(stepAditionalInformation(3))}
          />
        </div>
        <br />
        <button type="submit" className="btn btn-primary step__bold">
          Siguiente
        </button>
      </form>
    </>
  );
};
