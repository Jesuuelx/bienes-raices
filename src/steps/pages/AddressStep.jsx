import { useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { useDispatch } from "react-redux";
import { useState } from "react";
import { stepAddress } from "../../store/steps/stepsSlice";

const formValidations = {
  ciudad: [(value) => value.length >= 1, "El campo es obligatorio."],
  via: [(value) => value.length >= 1, "El campo es obligatorio."],
  nro1: [(value) => value.length >= 1, "El campo es obligatorio."],
  nro2: [(value) => value.length >= 1, "El campo es obligatorio."],
  nro3: [(value) => value.length >= 1, "El campo es obligatorio."],
};

const formData = {
  ciudad: "",
  via: "",
  nro1: "",
  nro2: "",
  nro3: "",
};

export const AddressStep = () => {
  const [submmited, setSubmmited] = useState(false);

  const dispatch = useDispatch();

  const navigate = useNavigate();

  const {
    ciudad,
    onInputChange,
    onResetForm,
    isFormValid,
    ciudadValid,
    via,
    viaValid,
    nro1,
    nro1Valid,
    nro2,
    nro2Valid,
    nro3,
    nro3Valid,
  } = useForm(formData, formValidations);

  const onSubmit = (e) => {
    e.preventDefault();
    setSubmmited(true);
    if (!isFormValid) return;
    dispatch(stepAddress({ address:`${ciudad} ${via} ${nro1} ${nro2} ${nro3}` }));
    navigate("/datos-piso", {
      replace: true,
    });
    onResetForm();
  };

  return (
    <>
      <h1 className="step__title">
        Paso 3 de 5: <br />{" "}
        <span className="step__intro-title"> Direccion *</span>
      </h1>
      <h2 className="mb-10 step__solicita">
        Â¿A que correo quieres que nos comuniquemos contigo?
      </h2>
      <hr />

      <form onSubmit={onSubmit} className="step__form-direccion">
        <div>
          <h4 className="mb-5"> Ciudad: </h4>
          <input
            type="text"
            placeholder="Ciudad"
            onChange={onInputChange}
            value={ciudad}
            name="ciudad"
            className="step__input-direccion"
            autoComplete="off"
          />
             {submmited && <p className="step__alert-error-direccion"> {ciudadValid} </p>}
        </div>
        <div>
          <h4 className="mb-5"> Via: </h4>
          <input
            type="text"
            placeholder="Calle/Carrera"
            onChange={onInputChange}
            value={via}
            name="via"
            className="step__input-direccion"
            autoComplete="off"
          />
             {submmited && <p className="step__alert-error-direccion"> {viaValid} </p>}
        </div>

        <div>
          <h4 className="mb-5"> Numero </h4>
          <input
            type="number"
            onChange={onInputChange}
            value={nro1}
            name="nro1"
            className="step__input-direccion"
            autoComplete="off"
          />
             {submmited && <p className="step__alert-error-direccion"> {nro1Valid} </p>}
        </div>
        <div>
          <h4 className="mb-5"> # </h4>
          <input
            type="number"
            onChange={onInputChange}
            value={nro2}
            name="nro2"
            className="step__input-direccion"
            autoComplete="off"
          />
             {submmited && <p className="step__alert-error-direccion"> {nro2Valid} </p>}
        </div>
        <div>
          <h4 className="mb-5"> - </h4>

          <input
            type="number"
            onChange={onInputChange}
            value={nro3}
            name="nro3"
            className="step__input-direccion"
            autoComplete="off"
          />
             {submmited && <p className="step__alert-error-direccion"> {nro3Valid} </p>}
        </div>
        <br />
        <button type="submit" className="btn btn-primary step__bold">Siguiente</button>
      </form>
    </>
  );
};
