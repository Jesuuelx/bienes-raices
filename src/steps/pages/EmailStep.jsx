import { useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { useDispatch } from "react-redux";
import { useState } from "react";
import { stepEmail } from "../../store/steps/stepsSlice";

const formValidations = {
  email: [
    (value) =>
      value.includes("@") &&
      !value.includes("@@") &&
      !value.includes("(") &&
      !value.includes(")") &&
      !value.includes(",") &&
      !value.includes("<") &&
      !value.includes(">") &&
      !value.includes(":") &&
      !value.includes("[") &&
      !value.includes("]") &&
      !value.includes(" ") &&
      !value.includes("%") &&
      !value.includes("&") &&
      !value.includes(";"),
    "Introduce un correo electronico valido",
  ],
};

const formData = {
  email: "",
};

export const EmailStep = () => {
  const [submmited, setSubmmited] = useState(false);

  const dispatch = useDispatch();

  const navigate = useNavigate();

  const { email, onInputChange, onResetForm, isFormValid, emailValid } =
    useForm(formData, formValidations);

  const onSubmit = (e) => {
    e.preventDefault();
    setSubmmited(true);
    if (!isFormValid) return;
    dispatch(stepEmail({ email: email }));
    navigate("/direccion", { replace: true });
    onResetForm();
  };

  return (
    <>
      <h1 className="step__title">Paso 2 de 5:<br />{" "}
        <span className="step__intro-title"> Correo Electronicó *</span>
      </h1>
      <h2 className="mb-10 step__solicita">
        ¿A que correo quieres que nos comuniquemos contigo?
      </h2>
      <hr />
      <h4 className="mb-5"> Correo: </h4>

      <form onSubmit={onSubmit} className="step__form"
      aria-label="step-two-form">
        <input
          type="text"
          placeholder="Ingresa tu nombre"
          onChange={onInputChange}
          value={email}
          name="email"
          className="step__input"
          autoComplete="off"
        />
        {submmited && <p className="step__alert-error"> {emailValid} </p>}
        <br />
        <button type="submit" className="btn btn-primary step__bold">
          Siguiente
        </button>
      </form>
    </>
  );
};
