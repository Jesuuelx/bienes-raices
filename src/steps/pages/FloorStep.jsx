import { useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { useDispatch } from "react-redux";
import { useState } from "react";
import { stepFloor } from "../../store/steps/stepsSlice";

const formValidations = {
  floor: [(value) => value.length >= 1 &&  value < 51 && value > 0 , "Debe introducir un valor valido ([1,  50])"],
};

const formData = {
  floor: "",
};

export const FloorStep = () => {
  const [submmited, setSubmmited] = useState(false);

  const dispatch = useDispatch();

  const navigate = useNavigate();

  const {
    floor,
    onInputChange,
    onResetForm,
    isFormValid,
    floorValid,
  } = useForm(formData, formValidations);

  const onSubmit = (e) => {
    e.preventDefault();
    setSubmmited(true);
    if (!isFormValid) return;
    dispatch(stepFloor({ floor: floor + 'piso' }));
      navigate('/caracteristicas-adicional', {replace:true});
    onResetForm();
  };

  return (
    <>
      <h1 className="step__title">
        Paso 4 de 5: <br />{" "}
        <span className="step__intro-title"> Piso de tu apartamento*</span>
      </h1>
      <hr />
      <h4 className="mb-5"> Piso: </h4>

      <form onSubmit={onSubmit} className="step__form">
        <input
          type="number"
          placeholder="Ingresa tu nombre"
          onChange={onInputChange}
          value={floor}
          name="floor"
          className="step__input"
          autoComplete="off"
        />

        {submmited && <p className="step__alert-error"> {floorValid} </p>}
        <br />
        <button type="submit" className="btn btn-primary step__bold">
          Siguiente
        </button>
      </form>
    </>
  );
};
