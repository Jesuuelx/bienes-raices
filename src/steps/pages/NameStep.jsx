import { useForm } from "../../hooks/useForm";
import { useDispatch } from "react-redux";
import { stepDisplayName } from "../../store/steps/stepsSlice";
import { useState } from "react";
import { useNavigate } from "react-router-dom";

const formData = {
  displayName: "",
};

const formValidations = {
  displayName: [(value) => value.length >= 1, "El campo es obligatorio."],
};
export const NameStep = ({ stepOne }) => {
  const [submmited, setSubmmited] = useState(false);

  const dispatch = useDispatch();

  const navigate = useNavigate();

  const {
    displayName,
    onInputChange,
    onResetForm,
    isFormValid,
    displayNameValid,
  } = useForm(formData, formValidations);

  const onSubmit = (e) => {
    e.preventDefault();
    setSubmmited(true);
    if (!isFormValid) return;
    dispatch(stepDisplayName({ name: displayName }));
    navigate(stepOne.pathTo, { replace: true });
    onResetForm();
  };

  return (
    <>
      <h1 className="step__title">
        Compramos <br />{" "}
        <span className="step__intro-title"> tu apartamento en 10 dias *</span>
      </h1>
      <h2 className="mb-10 step__solicita">Â¡Solicita una oferta gratis!</h2>
      <hr />
      <h3 className="mb-5 step__subtitle">Paso 1 de 5: Datos de registro</h3>
      <h4 className="mb-5">{stepOne.description}</h4>

      <form
        onSubmit={onSubmit}
        className="step__form"
        aria-label="step-one-btn"
      >
        <p className="step__bold">{stepOne.name}</p>
        <input
          type="text"
          onChange={onInputChange}
          value={displayName}
          name="displayName"
          className="step__input"
          autoComplete="off"
        />
        {submmited && <p className="step__alert-error"> {displayNameValid} </p>}
        <br />
        <button type="submit" className="btn btn-primary step__bold">
          Siguiente
        </button>
      </form>
    </>
  );
};
