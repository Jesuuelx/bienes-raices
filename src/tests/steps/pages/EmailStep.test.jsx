import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { configureStore } from "@reduxjs/toolkit";
import { stepsSlice } from "../../../store/steps/stepsSlice";
import { EmailStep } from "../../../steps/pages";

const store = configureStore({
  reducer: {
    steps: stepsSlice.reducer,
  },
  /*    preloadedState: {
          steps:
      } */
});


const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("Test in <EmailStep />", () => {
  
    beforeEach(() => jest.clearAllMocks());
    test("should make to match with Snapchot", () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <EmailStep />
        </MemoryRouter>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  test("should show a copy at h2", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <EmailStep />
        </MemoryRouter>
      </Provider>
    );

    expect(screen.getByRole("heading", { level: 2 }).innerHTML).toBe('Â¿A que correo quieres que nos comuniquemos contigo?'
    );
  });

  test('should press onSubmit and navigate to /address', () => { 

   render( <Provider store={store}>
    <MemoryRouter initialEntries={['/correo-usuario']}>
      <EmailStep />
    </MemoryRouter>
  </Provider>
   )
const input = screen.getByRole("textbox");
fireEvent.change(input, {
  target: { name: "email", value: "jesus@hotmail.com" },
});

const stepBtn = screen.getByLabelText("step-two-form");
fireEvent.submit(stepBtn);

expect( mockedUseNavigate ).toBeCalledWith('/direccion', {"replace": true});

   })

});
