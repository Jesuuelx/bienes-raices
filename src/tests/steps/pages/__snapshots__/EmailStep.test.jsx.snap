// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <EmailStep /> should make to match with Snapchot 1`] = `
<div>
  <h1
    class="step__title"
  >
    Paso 2 de 5:
    <br />
     
    <span
      class="step__intro-title"
    >
       Correo Electronicó *
    </span>
  </h1>
  <h2
    class="mb-10 step__solicita"
  >
    ¿A que correo quieres que nos comuniquemos contigo?
  </h2>
  <hr />
  <h4
    class="mb-5"
  >
     Correo: 
  </h4>
  <form
    aria-label="step-two-form"
    class="step__form"
  >
    <input
      autocomplete="off"
      class="step__input"
      name="email"
      placeholder="Ingresa tu nombre"
      type="text"
      value=""
    />
    <br />
    <button
      class="btn btn-primary step__bold"
      type="submit"
    >
      Siguiente
    </button>
  </form>
</div>
`;
